### PR TITLE
Add logic for saving a single imported interaction

### PR DESCRIPTION
**Pull request by reupen**
### Description of change

This adds the logic for saving a single imported interaction in the import interactions tool.

This will be used when the full logic for completing the import is added.

### Checklist

* [ ] Has a new newsfragment been created? Check [changelog/README.rst](https://github.com/uktrade/data-hub-leeloo/blob/master/changelog/README.rst) for instructions
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/interaction/admin_csv_import/row_form.py [hunk 1]
```diff
@@ -205,6 +206,35 @@ def full_clean(self):
             for field, errors in normalised_errors.items():
                 self._add_serializer_error(field, errors)
 
+    @atomic
+    def save(self, user, source):
+        """Creates an interaction from the cleaned data."""
+        serializer_data = self.cleaned_data_as_serializer_dict()
```
Comment (currycoder) [inline datahub/interaction/admin_csv_import/row_form.py]: Question:  Why can't you use the `serializer.save()` method with pertinent overrides instead?

File: datahub/interaction/test/admin_csv_import/test_row_form.py [hunk 2]
```diff
@@ -776,6 +777,132 @@ def test_cleaned_data_as_serializer_dict_for_service_delivery(self):
...
+                (adviser_1, adviser_1.dit_team),
+                (adviser_2, adviser_2.dit_team),
+            ),
+        )
+
+        assert actual_advisers_and_teams == expected_advisers_and_teams
+
+    def test_save_rolls_back_on_error(self, monkeypatch):
+        """Test that save() rolls back if there's an error."""
+        monkeypatch.setattr(
```
Comment (currycoder) [inline datahub/interaction/test/admin_csv_import/test_row_form.py]: Nice test!

File: datahub/interaction/test/admin_csv_import/test_row_form.py [hunk 3]
```diff
@@ -776,6 +777,132 @@ def test_cleaned_data_as_serializer_dict_for_service_delivery(self):
...
+            'date': '02/03/2018',
+            'adviser_1': adviser.name,
+            'contact_email': contact.email,
+            'service': service.name,
+            'communication_channel': communication_channel.name,
+        }
+        form = InteractionCSVRowForm(data=data)
+
+        assert form.is_valid()
+        interaction = form.save(user, source=source)
```
Comment (currycoder) [inline datahub/interaction/test/admin_csv_import/test_row_form.py]: Rather than trusting the return value, could you look up the interaction independently to ensure that it was persisted before making assertions?

Comment (reupen) [reply datahub/interaction/test/admin_csv_import/test_row_form.py]: I'll add an `interaction.refresh_from_db()` :+1:

Comment (currycoder) [reply datahub/interaction/test/admin_csv_import/test_row_form.py]: :+1:

File: datahub/interaction/admin_csv_import/row_form.py (hunk 1 above)
Comment (reupen) [reply datahub/interaction/admin_csv_import/row_form.py]: Hmm, I could call `InteractionSerializer.create()` directly. Could be a bit fragile though...

Comment (currycoder) [reply datahub/interaction/admin_csv_import/row_form.py]: Sure!  I was mainly hoping to open a discussion about it, my next PR for calendar processing makes use of `serializer.save()` and things are working as I expect.  I suppose you are concerned about future changes to serializer field validation etc for the API which could prove to be problematic?

Comment (reupen) [reply datahub/interaction/admin_csv_import/row_form.py]: Sort of. To use `InteractionSerializer.save()`, I thought I'd have to reinstate #1633 (or put the data in the format used for an API request). And, run full serialiser validation and pass read-only etc. fields separately as keyword arguments to the `InteractionSerializer.save()` call.

(Alternatively, `InteractionSerializer.create()` could be used which would make things much simpler, but seems a bit hacky.)

So, in summary, not against using `InteractionSerializer.save()` but we would need to make more changes as this would be the first time calling `is_valid()` on the serialiser (so we would need to change the earlier validation stage of the import to do that as well).

Comment (currycoder) [reply datahub/interaction/admin_csv_import/row_form.py]: Yep makes sense.  I'm easy either way really - I think the code that you have here is neat and understandable.  I think perhaps we just need to document on the serialiser that there is some usage for the validation (or full usage) outside of the API layer, as this may not be obvious for somebody making modifications to it that is unaware.

File: datahub/interaction/test/admin_csv_import/test_row_form.py [hunk 4]
```diff
@@ -776,6 +777,134 @@ def test_cleaned_data_as_serializer_dict_for_service_delivery(self):
...
+        }
+        form = InteractionCSVRowForm(data=data)
+
+        assert form.is_valid()
+
+        with pytest.raises(ValueError):
+            form.save(user, source=source)
+
+        assert not Interaction.objects.exists()
+
```
Comment (marcofucci) [inline datahub/interaction/test/admin_csv_import/test_row_form.py]: It would be good to have a test for when a contact is not matched. My understanding is that `is_valid` would silently return but the save would fail?

Comment (reupen) [reply datahub/interaction/test/admin_csv_import/test_row_form.py]: I've added a test and made it a little bit more robust for that situation.

File: datahub/interaction/admin_csv_import/row_form.py (hunk 1 above)
Comment (reupen) [reply datahub/interaction/admin_csv_import/row_form.py]: Yes, it would be good to make things as clear as possible.

There was a small note above the validators on the serialiser, but I've added a more detailed one in the main docstring for `InteractionSerializer`.
